internal/expr/parser.go

```diff
@@ -716,30 +735,139 @@ func (p *Parser) parseOutputExpr() (*outputExpr, bool, error) {
 	return nil, false, nil
 }
 
-// parseInputExpr parses an input expression of the form "$Type.name".
-func (p *Parser) parseInputExpr() (expression, bool, error) {
+// parseSliceInputExpr parses an input expression of the form "$Type[:]".
+func (p *Parser) parseSliceInputExpr() (*sliceInputExpr, bool, error) {
 	cp := p.save()
 	if !p.skipByte('$') {
 		return nil, false, nil
 	}
 
-	// Case 1: Slice range, "Type[:]".
 	if st, ok, err := p.parseSliceAccessor(); err != nil {
+		cp.restore()
 		return nil, false, err
 	} else if ok {
 		return &sliceInputExpr{sliceTypeName: st, raw: p.input[cp.pos:p.pos]}, true, nil
 	}
 
-	// Case 2: Struct or map, "Type.something".
-	if ma, ok, err := p.parseTypeAndMember(); ok {
-		if ma.memberName == "*" {
-			return nil, false, errorAt(fmt.Errorf("asterisk not allowed in input expression %q", "$"+ma.String()), cp.lineNum, cp.colNum(), p.input)
+	cp.restore()
+	return nil, false, nil
+}
+
+// parseMemberInputExpr parses an input expression of the form "$Type.member".
+func (p *Parser) parseMemberInputExpr() (*memberInputExpr, bool, error) {
+	cp := p.save()
+	ma, ok, err := p.parseInputMemberAccessor()
+	if !ok {
+		cp.restore()
+		return nil, false, err
+	}
+	if ma.memberName == "*" {
+		cp.restore()
+		return nil, false, errorAt(fmt.Errorf("invalid asterisk input placement %q", "$"+ma.String()), cp.lineNum, cp.colNum(), p.input)
+	}
+	return &memberInputExpr{ma: ma, raw: p.input[cp.pos:p.pos]}, true, nil
+}
+
+// parseAsteriskInputExpr parses an INSERT statement input expression where
+// SQLair generates the columns.
+// It is of the form "(*) VALUES ($Type.*, $Type.member,...)".
+func (p *Parser) parseAsteriskInputExpr() (*asteriskInputExpr, bool, error) {
+	cp := p.save()
+
+	parenCol := p.pos + 1
```

Comment:
This will not work because the column could be on a separate line. You have to do cp.colNum() or p.colNum() and increment it when you are sure there was a parentheses.